Extract Google Drive API

Kordishal opened the pull request:
I want to extract the Google Drive API parts from the client class to better distinguish the two. This should help make it a bit clearer and clean up the Client class a bit which has a ton of functions.

The goal is to improve the way the Google Drive API is used for the useful features. And potentially to have the option to load pygsheets without granting access to the Google Drive API, as this may be a security issue.

Currently only the export function is changed. All the functionality is moved from the Spreadsheet & Worksheet classes to the DriveAPIWrapper class. 

The implementation is changed that always all sheets will be exported, but to have the option to export specific sheets into CSV & TSV.

Adds support for TSV & HTML export.
Added a lot of documentation.
Added a test.

Need to still update the documentation and add an additional test. 

What do you think?

nithinmurali commented:
> The goal is to improve the way the Google Drive API is used for the useful features. And potentially to have the option to load pygsheets without granting access to the Google Drive API, as this may be a security issue.

The major reason drive api acess is not optional is because it it used for mapping ssheet name to ssheet key, which i think is the most used feature. But yeah if someone is not using those, we can give an option. 

If we are writing a wraper for Drive Api shouldn't we analogously write an wrapper for Sheets api too ? Especially as you are placing the wrapper in api/.

> I want to extract the Google Drive API parts from the client class to better distinguish the two. This should help make it a bit clearer and clean up the Client class a bit which has a ton of functions.

I am not really sure is this will clean the client ? because we still have to define all the functions in the client. As currently client is the only interface between the api and models.

Kordishal commented:
>The major reason drive api acess is not optional is because it it used for mapping ssheet name to ssheet key, which i think is the most used feature. But yeah if someone is not using those, we can give an option.

Yes by default Drive API will be used. The drive API is necessary for exporting, deleting, listing sheets, all permission related stuff (and if they are implemented comments). So it would be hard to separate and making it clear which functionally works without Drive API access.

Another difference is, that the Drive API does not have a Quota as limiting as Sheets API. It is like 1'000'000 API calls a day. So if we want to implement #143 we only need to count API calls to the Sheets API.

>  If we are writing a wraper for Drive Api shouldn't we analogously write an wrapper for Sheets api too ? Especially as you are placing the wrapper in api/.

That is the plan yes. But I think doing both at the same time is not a good idea.

> I am not really sure is this will clean the client ? because we still have to define all the functions in the client. As currently client is the only interface between the api and models.

Basically the client would have a drive attribute and a service/sheetapi attribute which would expose the functions defined there. These would be used through the client by the models. But the user does not need most of them, unless they want to use unsupported/unimplemented functionality. 

The client only really needs to expose `open`, `open_by_key`, `open_by_url`, `open_all`, `list_ssheets` and `create`. All other functionality is offered by the sheets/worksheets themselves and the functions in client are only go betweens which implement the requests. (The exception would be `get_range` which has the potential to offer direct data access without having to load an entire spreadsheet. This would be useful for people who only use google sheets as read only data source (like configs)).

If we have a SheetsAPIWrapper all `sh_append` functions would be implemented there and accessed by the Models through `self.client.sheetapi` (which would be an instance of SheetsAPIWrapper).

This way the client would be used as a starting point into the spreadsheets, but then the models are used for actual operations on the data. 

And if someone really needs to use a functionality which is not (yet) offered by the models they can do so by using the `client.sheetapi.service` which would expose the full googleapiclient API.

The primary advantage would be to have an easier time to maintain this package since the functionalities would be clearly separated. Or do you think a lot of people use `sh_append` directly?

Kordishal commented:
See [here](https://github.com/Kordishal/pygsheets/blob/drive-file-api-wrapper/pygsheets/api_wrappers/drive_api_wrapper.py) for a more complete version of the drive API. There I ran into the problem, that I made a change which affected too many systems and I wanted first to ask you want you think of these changes.

nithinmurali commented:
As such I like what you are proposing. I have planed in the beginning of the project to make switching of back-ends easier if needed. I think this would also be a move in that direction. 
So if you are planning to work on this go ahead. But i am not really sure if this is worth the effort, given that you would need to do a complete revamp to pull this off.

Kordishal commented:
Yes it would be a bit of work, but I think I can just extract the Drive API first and the Sheet API later. 
Should I run out of time to invest in the project and you already would have some advantage in maintaining the package. Which is really great to use btw!

Kordishal commented:
With this the permission system is reworked. What's left for the Drive API is the list, delete and move functionalities.

The permissions are kept fairly simple. You can see all the permissions on a spreadsheet. Remove one by of them and add a new one with an email/domain or make it public. The permissions are not stored locally as I feel that there is no point in doing that. Permissions have no effect on local processing.

Kordishal commented:
Finished implementation of all the Drive API functionality in DriveAPIWrapper which is used in pygsheets. Expanded some of it or made slight adjustments on how it works. 

Major Changes:

- Replaced `list_ssheets()` with `spreadsheet_ids()` & `spreadsheet_titles()`. Should someone require both for something they will have to call `drive.spreadsheet_metadata()`
- Changed `open_all(query)` to now return a list of all spreadsheets. The query is added to the mimeType query and must adhere to the query documentation.
- Added `drive.move_file` to make `create` easier. 
- Moved `delete` to the spreadsheet model. Now a simple call to `spreadsheet.delete()` is necessary to delete a spreadsheet. This has required some refactoring in the test suite. 
- Added `open_as_json` to `Client` and removed the json option from `open_by_key`.
- Behaviour of `open`, `open_by_key` and `open_by_url` should be unchanged.

All of these changes should be documented in the relevant functions.

Need to review and update the documentation and figure out how to integrate the DriveAPIWrapper documentation into everything. How would you say is the best way to integrate it? Have a section next to the Models section on API Wrapper?

Should I add a test suite for the DriveAPIWrapper? The functionality is tested through the model test suits and I feel this would mostly be duplication.

The only problem is that the offline tests are failing on Travis due to config being None and I am  not sure why that is. All online tests succeed.

Aside that and documentation this should be ready to review and then merge if everything is ok.

For the SheetAPIWrapper I will create a separate PR.

nithinmurali commented:
> Need to review and update the documentation and figure out how to integrate the DriveAPIWrapper documentation into everything. How would you say is the best way to integrate it? Have a section next to the Models section on API Wrapper?

Yup that would do

> Should I add a test suite for the DriveAPIWrapper? The functionality is tested through the model test suits and I feel this would mostly be duplication.

separate tests are not necessary , as they are already getting tested.

> The only problem is that the offline tests are failing on Travis due to config being None and I am not sure why that is. All online tests succeed.

I will take a look.

> Aside that and documentation this should be ready to review and then merge if everything is ok.

cool, will begin the review once i am free

nithinmurali commented:
Currently based on how you have implemented, all the requests are executed at their corresponding functions. But this would kill the batch mode functionality and also the retry requests.

nithinmurali commented on pygsheets/client.py:
```diff
@@ -74,152 +75,81 @@ def __init__(self, oauth, http_client=None, retries=1, no_cache=False):
-        return results
+    def spreadsheet_titles(self, query=None):
+        """A list of all the titles of spreadsheets present in the users drive or TeamDrive."""
+        return [x['name'] for x in self.drive.spreadsheet_metadata(query)]
```
why not cache the spreadsheet metadatas ? like it was done using  `_spreadsheets`? As there is a less chance for the spreadsheet's metadata's (at-least title and id to be changed) while script is running, and if its changed then update the cache (it is implemented in previous code).

Kordishal replied:
I changed this because the Drive API does not count towards the same API limitations as Sheet API calls. So there is no need to limit the number of calls. And making the request does not take long at all. And is only very rarely called in normal operation. 

This way ensures that the data is always up to date. In the old implementation it was up to the user to realize when to update the metadata. (it would try to load new data when a sheet was not found or when no sheets were loaded yet.) 

So overall I do not really see the advantage of the old implementation.

nithinmurali replied:
Ahh, its Cool then.

Kordishal commented on pygsheets/spreadsheet.py:
```diff
@@ -272,35 +271,40 @@ def find(self, pattern, **kwargs):
 
-    def remove_permissions(self, addr):
-        """Remove user from permissions list.
+    def remove_permission(self, permission):
```
This makes sense. However I think it is enough to offer deletion by email/domain without id. Users will usually not know the ID of a permission. And if they load the permission from the list it makes hardly a difference if they extract the email/domain or id. What do you think?

nithinmurali commented:
> Currently based on how you have implemented, all the requests are executed at their corresponding functions. But this would kill the batch mode functionality and also the retry requests.

About this?

Kordishal commented:
I can't find the documentation for `self.service.new_batch_http_request(callback=callback)`. Where did you find this?

Because at the moment I only see this function which then makes all the requests in succession. I guess it makes all the requests and then uses the callback to determine what to do with the response? In this case just printing put the response?

As for reimplementing the batch mode for Drive API funtionality. Is this actually needed? The problem I see is, that all the Drive API calls are crucial to get any data in the first place. So all subsequent functions used may not use any data until the batch request is called. But in Spreadsheet there are a number of functions which manipulate/use data before making a request. 

Is the idea that people who use the batch mode are aware of this when deciding on the calls made? Or am I missing something?

Implementing batch_mode should be possible, but will require an additional param on each function in the Drive API. If you think batch mode is absolutely essential to include Drive API calls then I will do that. But Drive API calls are primarily for setup and do not use a lot of time.

nithinmurali commented:
The basic idea of batch mode was when there are many update requests, we could merge them and send as a single update request saving both time and number of requests. Though the effecient merging requests is not yet implemented.  
But I see that drive service dosen't do any update operation, so i think we don't need batch request options in drive.

nithinmurali commented:
If you have no more changes, then i will merge this.

Kordishal commented:
Nope, working on some doc updates, but since there is more than just updates for this PR I will be making a separate PR for that.

Kordishal commented:
So I found the bug which made the Travis build fail. Now only the 2.6 build fails. Which is not supported by the Google API Client anyway: https://github.com/google/google-api-python-client

So should I remove that? And add 3.6/3.7 instead?

nithinmurali commented:
> So should I remove that? And add 3.6/3.7 instead?

Yeah! please do that.

Kordishal commented:
So I can either get the 2.6 build to work or the 3.x builds. The reason is the requirements installation.

2.6 needs enum to be installed while the other versions may not do that. Do you know of a way to make Travis CI do that? Didn't find something in the docs yet.

nithinmurali commented:
Once you are done with the PR, remove the [WIP] from heading, i will test and merge.

nithinmurali commented:
Thanks @Kordishal for this PR. This is a major improvement for pygsheets.
